File: internal/ui/workspace.go
package ui

import (
	"bytes"
	"fmt"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/dialog"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"

	"github.com/flacyak/uno/internal/document"
	"github.com/flacyak/uno/internal/ingest"
	"github.com/flacyak/uno/internal/pattern"
	"github.com/flacyak/uno/internal/sheet"
)

// A workspace is one open file: one tab, one sheet, one grid, one .uno (I-3).
// It is also the blast radius. The edit log hangs off the sheet this struct
// owns, so a change made in one file can never reach another.
type workspace struct {
	sheet  *sheet.Sheet
	table  *grid
	editor *widget.Entry
	tab    *container.TabItem

	// inline is the editor that appears in the grid itself, inlineIn the cell
	// holding it, and editing whether it is in the grid at all. Which cell it is
	// in is not a fourth piece of state: only the selected cell is ever edited,
	// so active already says.
	inline   *inlineEntry
	inlineIn *cell
	editing  bool

	// proposal is what the recogniser last found here, and nil when it found
	// nothing or the log has moved on. The bar that asks about it belongs to the
	// shell, because only the selected tab is being looked at; what the question
	// is about stays here, so nothing about one file reaches another (I-3).
	proposal *pattern.Proposal

	// formulaRefs says which .unof each bound column came from, so edit beside a
	// name can find it again. It is a convenience and never a dependency: the
	// expression itself is in the log, so a file whose references resolve to
	// nothing still computes (I-4).
	formulaRefs map[int]string

	// dismissed remembers, per column, the program the person said no to. Keyed
	// by the program and not only by the column, so refusing one offer does not
	// silence a column for the session: a better question about it is a
	// different question.
	dismissed map[int]string

	// raw is the bytes uno was handed, kept for as long as the workspace is
	// open. They are authoritative (I-4): the .uno stores them verbatim, and
	// replay rebuilds the sheet from them rather than from a rewritten copy.
	raw []byte

	// name is what the tab says: the .uno's name once saved, and the source
	// file's before that. It is not sheet.Name, which stays the name of the file
	// the bytes came from so ingest keeps choosing the same decoder for them.
	name string

	// path is the .uno this workspace writes to. Empty until the first save,
	// which is what makes Ctrl+S fall through to Save As.
	path string

	// manifest carries provenance across saves: the source name it was opened
	// from, and the time it was first written.
	manifest document.Manifest

	// extra holds entries a newer uno wrote that this build did not recognise,
	// so saving carries them through rather than dropping them.
	extra map[string][]byte

	active document.Cell
	saving bool

	// savedLog is the log as it was written to the file on disk, and nil for a
	// workspace that has never been saved. Holding it is what lets undo take the
	// dot back off: the workspace is clean whenever what it holds is what was
	// written, however it got back there.
	savedLog []sheet.Edit
}

// dirty reports whether this workspace holds changes no file has.
func (w *workspace) dirty() bool {
	return w.sheet != nil && !w.sheet.LogEquals(w.savedLog)
}

// newWorkspace builds an empty workspace showing the drop target. It gets its
// real name when a file lands in it.
func (s *Shell) newWorkspace() *workspace {
	w := &workspace{name: untitled(len(s.tabs.Items) + 1)}
	w.tab = container.NewTabItem(w.name, s.emptyState())
	s.byTab[w.tab] = w
	return w
}

func untitled(n int) string { return fmt.Sprintf("Untitled %d", n) }

// fill swaps a loaded sheet into this workspace and replaces the drop target
// with the grid. Nothing outside the workspace sees the sheet until this
// returns, so a failed load leaves the previous content untouched.
//
// The caller sets name, path, manifest, extra and active first: this builds
// what those describe.
func (s *Shell) fill(w *workspace, sh *sheet.Sheet, raw []byte) {
	w.sheet = sh
	w.raw = raw
	w.editor = s.newEditor(w)

	// The grid's cells reach for the inline editor as they are built, so it has
	// to exist before the table does.
	w.inline, w.inlineIn, w.editing = s.newInline(w), nil, false
	w.table = s.newTable(w)
	w.proposal, w.dismissed = nil, map[int]string{}

	w.table.OnSelected = func(id widget.TableCellID) {
		if id.Row < 0 || id.Col < 0 {
			return // a header is not a cell anyone edits
		}
		w.active = document.Cell{Row: id.Row, Col: id.Col}
		w.showActive()
		s.refreshStatus()
	}

	// The editor bar sits above the grid, which is the one place a value too
	// wide for its column is still readable in full.
	w.tab.Content = container.NewBorder(w.editor, nil, nil, nil, w.table)
	w.tab.Text = w.name

	// Selecting through the table rather than assigning w.active is what puts
	// the highlight, the editor and the cell reference in the same state a click
	// would leave them in.
	if sh.Rows() > 0 && sh.Cols() > 0 {
		w.table.Select(widget.TableCellID{
			Row: clampIndex(w.active.Row, sh.Rows()),
			Col: clampIndex(w.active.Col, sh.Cols()),
		})

		// Select scrolls to what it selected, and here it is doing that
		// arithmetic against a table with no size yet, so the offset it lands on
		// means nothing. A freshly opened workspace shows the start of its file.
		// Coming back to where the view sat is a separate promise, and it needs
		// a scroll position stored as a row identity rather than as an offset.
		w.table.ScrollToTop()
		w.table.ScrollToLeading()

		// A file that has just opened has a cell chosen and nothing else worth
		// typing into, so the grid takes the keyboard: without this the arrows
		// and Enter do nothing until something has been clicked, which makes the
		// keyboard a thing you reach by using the mouse first.
		s.focus(w.table)
	}

	// A .uno reopened part-way through fixing a column arrives with the examples
	// already in its log, and the question is as worth asking on Monday as it
	// was on Friday.
	s.rescan(w)
}

// newEditor is the cell editor: one field, not a widget per cell. The grid keeps
// recycling plain labels, so its update closure stays the allocation-free read
// that I-1 promises however much of the sheet is being edited.
func (s *Shell) newEditor(w *workspace) *widget.Entry {
	e := widget.NewEntry()
	e.SetPlaceHolder("Select a cell, then type here and press Enter")
	e.OnSubmitted = func(text string) { s.commit(w, text) }
	return e
}

// commit writes what was typed into the selected cell, through sheet.Set so the
// change is logged as it is made. Retyping the value already there is not an
// edit and must not add a line to the log.
func (s *Shell) commit(w *workspace, text string) {
	if w.sheet == nil || text == w.sheet.Raw(w.active.Row, w.active.Col) {
		return
	}

	before := w.sheet.Columns[w.active.Col]
	if err := w.sheet.Set(w.active.Row, w.active.Col, text); err != nil {
		dialog.ShowError(err, s.win)
		return
	}

	// One cell changed, so one cell is redrawn. A column that changed kind
	// changed its badge as well, and that badge lives in the header.
	if w.sheet.Columns[w.active.Col] != before {
		w.table.Refresh()
	} else {
		w.table.RefreshItem(widget.TableCellID{Row: w.active.Row, Col: w.active.Col})
	}
	w.showActive()
	s.rescan(w) // this edit is the evidence the recogniser learns from
	s.refreshStatus()
}

// showActive points the editor bar at the selected cell. Every path that changes
// which cell that is, or what is in it, ends here, so the bar and the grid
// cannot drift apart.
//
// Raw and not Display: the bar is where a cell is edited, and what you edit is
// what the cell stores. The grid is the only thing bound to Display.
func (w *workspace) showActive() {
	w.editor.SetText(w.sheet.Raw(w.active.Row, w.active.Col))
}

// inlineEntry is the grid's cell editor. It is a plain Entry apart from Escape:
// Fyne's Entry ignores that key, and an editor opened by a stray click needs a
// way out that writes nothing.
type inlineEntry struct {
	widget.Entry
	cancel func()
}

func (s *Shell) newInline(w *workspace) *inlineEntry {
	e := &inlineEntry{cancel: func() { s.endEdit(w, false) }}
	e.ExtendBaseWidget(e)
	e.OnSubmitted = func(string) { s.endEdit(w, true) }
	return e
}

func (e *inlineEntry) TypedKey(k *fyne.KeyEvent) {
	if k.Name == fyne.KeyEscape {
		e.cancel()
		return
	}
	e.Entry.TypedKey(k)
}

// tapCell is a click on a square of the grid. The first click chooses the cell,
// which is all a click has ever done; a second click on the cell already chosen
// opens it for typing, so the fix happens where the value is rather than at the
// top of the window. Double-click would be the more familiar gesture and cannot
// be the one: Fyne delays every single tap on a double-tappable object while it
// waits to see whether a second one is coming, and that delay is on the plain
// selection click, which is the one uno does most.
func (s *Shell) tapCell(w *workspace, id widget.TableCellID) {
	if w.sheet == nil {
		return
	}
	if w.active == (document.Cell{Row: id.Row, Col: id.Col}) {
		s.beginEdit(w)
		return
	}

	// Clicking off a cell keeps what was typed into it, the way a spreadsheet
	// does. Selecting is what then repoints active, the bar and the status.
	s.endEdit(w, true)
	w.table.Select(id)

	// widget.Table focuses itself when it handles a tap, and it did not handle
	// this one. Without this, giving a cell its own click would quietly cost the
	// grid the keyboard.
	s.focus(w.table)
}

// beginEdit puts the inline editor into the selected cell, carrying the value
// already there so correcting one character is not retyping the whole field.
func (s *Shell) beginEdit(w *workspace) {
	if w.sheet == nil || w.editing || w.sheet.Rows() == 0 {
		return
	}
	w.editing = true
	w.inline.SetText(w.sheet.Raw(w.active.Row, w.active.Col))
	w.table.RefreshItem(widget.TableCellID{Row: w.active.Row, Col: w.active.Col})
	s.focus(w.inline)
}

// endEdit takes the inline editor back out of the grid. keep says whether what
// is in it is worth anything: Enter and a click on another cell keep it, Escape
// does not. Committing goes through the same commit the editor bar uses, so a
// value typed in the grid is logged exactly like one typed at the top.
func (s *Shell) endEdit(w *workspace, keep bool) {
	if !w.editing {
		return
	}
	text := w.inline.Text
	w.editing = false
	w.table.RefreshItem(widget.TableCellID{Row: w.active.Row, Col: w.active.Col})
	s.focus(w.table)

	if keep {
		s.commit(w, text)
	}
}

// focus guards the canvas, which a shell built without one does not have.
func (s *Shell) focus(o fyne.Focusable) {
	if c := s.win.Canvas(); c != nil {
		c.Focus(o)
	}
}

// undo drops the last operation and rebuilds the sheet by replaying what is left
// over the raw bytes.
//
// Rebuilding rather than putting the old value back is what makes undo survive a
// save: a file closed on Friday still undoes on Monday, because the raw bytes
// and the log travel in the .uno and nothing about undo was ever held in memory.
// It is also the only mechanism that stays right once one operation covers
// thousands of rows and has no old value to put back.
func (w *workspace) undo() error {
	edits := w.sheet.Edits()

	sh, err := ingest.Read(w.sheet.Name, bytes.NewReader(w.raw))
	if err != nil {
		return err
	}
	if err := sh.Replay(edits[:len(edits)-1]); err != nil {
		return err
	}

	// Only a rebuild that got all the way here replaces what is on screen.
	w.sheet = sh
	return nil
}

// document snapshots the workspace as a value the writer can take away. The raw
// bytes are immutable and everything else is copied here, so the deflate and the
// fsync run on a worker goroutine while the person carries on typing (I-7). The
// grid's dimensions are read here for the same reason: they are the one thing
// the manifest states that the writer cannot measure for itself.
func (w *workspace) document() *document.Document {
	m := w.manifest
	m.Sheet.Rows, m.Sheet.Cols = w.sheet.Rows(), w.sheet.Cols()

	return &document.Document{
		Manifest: m,
		Raw:      w.raw,
		State: document.State{
			Active:         w.active,
			ColumnFormulas: w.columnFormulas(),
		},
		Edits: w.sheet.Edits(),
		Extra: w.extra, // replaced wholesale on open, never written into
	}
}

// columnFormulas is the library reference for each bound column, in column
// order so two saves of an unchanged workspace produce the same bytes.
//
// A column with no reference is simply absent. That is the ordinary case for a
// file someone else bound and sent, and it is not a gap to be filled in: the
// expression is in the log, and the reference is only ever how the drawer finds
// the formula again on the machine it was written on.
func (w *workspace) columnFormulas() []document.ColumnFormula {
	if len(w.formulaRefs) == 0 {
		return nil
	}
	out := make([]document.ColumnFormula, 0, len(w.formulaRefs))
	for col := 0; col < w.sheet.Cols(); col++ {
		if ref, ok := w.formulaRefs[col]; ok {
			out = append(out, document.ColumnFormula{Col: col, Ref: ref})
		}
	}
	return out
}

func clampIndex(i, n int) int {
	if i < 0 || i >= n {
		return 0
	}
	return i
}

// emptyState is the dashed drop target. It is a standing affordance, not a
// hover state: the window API hands us the drop, not the drag-over, so there is
// no event to light a border up with while a file is held over the window.
func (s *Shell) emptyState() fyne.CanvasObject {
	title := widget.NewLabelWithStyle("Drop a data file here", fyne.TextAlignCenter,
		fyne.TextStyle{Bold: true})
	blurb := widget.NewLabelWithStyle(
		"uno reads it from disk and keeps it there. Nothing leaves this machine.",
		fyne.TextAlignCenter, fyne.TextStyle{})
	exts := widget.NewLabelWithStyle(".csv  .tsv  .json  .uno", fyne.TextAlignCenter,
		fyne.TextStyle{Monospace: true})

	btn := widget.NewButtonWithIcon("Open…", theme.FolderOpenIcon(), s.chooseFile)
	btn.Importance = widget.HighImportance

	return container.NewCenter(container.NewVBox(
		title,
		blurb,
		container.NewHBox(layout.NewSpacer(), btn, layout.NewSpacer()),
		exts,
	))
}

File: internal/desktop/src/renderer/workspace.ts
// One open workspace: its sources, which one is showing, and where it goes back to.
//
// A workspace is an engine and, for each source in it, the band of rows it
// sends. The files and the log live in the engine, so view and transform read
// the same rows the same way: transform unlocks editing and turns the
// recogniser on, and loads nothing. An edit is a message and a new generation,
// which is why applying a program to 50 million rows reaches the screen as fast
// as typing into one cell.
//
// It holds no widgets, which is the property that let the Go build test its
// shell without a window.

import { Band } from "@uno/grid/engine";
import type {
  Changed,
  Engine,
  FindRequest,
  Found,
  Link,
  Offer,
  SourceHandle,
  SourceRef,
} from "@uno/grid/engine";
import { NO_ROW, Op, editEquals } from "@uno/grid/sheet";
import type { Edit } from "@uno/grid/sheet";

import type { Cell, Rows } from "./grid/rows.ts";

/**
 * The most a saved workspace carries, all such sources together.
 *
 * A .uno points at the files it can name, so this bounds only what is left:
 * bytes with no file behind them, which the container has to copy or lose. On
 * the desktop every source comes from a path, so nothing counts against it.
 */
export const CARRY_LIMIT = 256 << 20;

export type Mode = "view" | "transform";

/** One source as the workspace shows it: a tab. */
export class Tab {
  /** The recogniser's question about this source, while it has one. */
  offer: Offer | null = null;
  /** Where the selection was when another tab was shown, so coming back finds it. */
  cell: Cell = { row: 0, col: 0 };

  /** The log as the engine last reported it, and as it was at the last save. */
  private edits: Edit[];
  private savedEdits: Edit[];

  constructor(
    readonly source: SourceHandle,
    readonly band: Band,
    /** The log the last save held, for a tab replacing one that was already
     * here. A tab that has just opened was saved with whatever it opened with. */
    saved?: readonly Edit[],
  ) {
    this.edits = source.opened.edits.slice();
    this.savedEdits = (saved ?? source.opened.edits).slice();
  }

  get id(): string {
    return this.source.id;
  }

  get name(): string {
    return this.source.opened.name;
  }

  /** The file behind this source, for a tab that points at one. */
  get link(): Link | undefined {
    return this.source.opened.link;
  }

  /** What is wrong with the file behind this source, if anything: it is gone,
   * or it is not the file the log was written against. */
  get trouble(): string | undefined {
    return this.link?.missing ?? this.link?.changed;
  }

  /** Whether there is a grid behind this tab at all. */
  get missing(): boolean {
    return this.link?.missing !== undefined;
  }

  /** The log the last save held, so a tab replacing this one keeps the same
   * idea of what is unsaved. */
  get savedLog(): readonly Edit[] {
    return this.savedEdits;
  }

  get edited(): number {
    return this.edits.length;
  }

  /**
   * dirty compares the log rather than counting it.
   *
   * Undo makes a count ambiguous: taking one edit back and making a different
   * one lands on the same number and a different sheet.
   */
  get dirty(): boolean {
    const a = this.edits;
    const b = this.savedEdits;
    return a.length !== b.length || !a.every((e, i) => editEquals(e, b[i]!));
  }

  landed(changed: Changed): void {
    this.edits.push(changed.edit);
    this.band.columns = changed.columns;
  }

  undone(changed: Changed): void {
    this.edits.pop();
    this.band.columns = changed.columns;
  }

  saved(): void {
    this.savedEdits = this.edits.slice();
  }
}

export class Workspace {
  /** Where a save goes without asking. Empty until the workspace has been saved
   * once, or was opened from a .uno. */
  path = "";

  /** Never saved: every open lands in view. */
  mode: Mode = "view";

  /** In the order they were added, which is the order the strip shows. */
  private tabs: Tab[] = [];
  private showing!: Tab;
  /** The sources the last save held, so adding or removing one is unsaved work too. */
  private savedSources: string[] = [];
  /**
   * Sources pointed at a different file since the last save.
   *
   * The log does not change when a source is relinked, and neither does the
   * list of sources, so nothing else here would notice. What changed is the
   * path the .uno on disk still holds, and leaving that unsaved is how somebody
   * finds the same missing file again tomorrow.
   */
  private readonly relinked = new Set<string>();

  private constructor(
    private readonly engine: Engine,
    /** Called when rows land or the index moves, so whoever draws can draw them. */
    private readonly changed: () => void,
    /** Called when a source's offer changes, with the tab it belongs to. */
    private readonly offered: (tab: Tab) => void,
  ) {}

  /**
   * open reads a file through an engine the workspace owns from here on.
   *
   * `savePath` is where Ctrl+S writes without asking. It only applies to a .uno,
   * since anything else has no workspace file to go back to, and a dropped .uno
   * passes "" so its first save asks.
   */
  static async open(
    ref: SourceRef,
    engine: Engine,
    savePath: string,
    changed: () => void,
    offered: (tab: Tab) => void,
  ): Promise<Workspace> {
    const w = new Workspace(engine, changed, offered);
    w.showing = await w.add(ref);
    // What it opened with is nothing to lose: the file is still where it was.
    // A source added after it is, until a save keeps it.
    w.savedSources = w.tabs.map((t) => t.id);
    if (ref.name.toLowerCase().endsWith(".uno")) w.path = savePath;
    return w;
  }

  /**
   * add puts a file's sources in the workspace, and returns the one to show:
   * the file just added, or the source a .uno was left on. What shows stays
   * the caller's decision, since the grid has to follow it.
   */
  async add(ref: SourceRef): Promise<Tab> {
    const { sources, showing } = await this.engine.open(ref);
    const added = sources.map((s) => this.tab(s));
    this.tabs.push(...added);
    return added.find((t) => t.id === showing) ?? added[0]!;
  }

  private tab(source: SourceHandle, saved?: readonly Edit[]): Tab {
    const t = new Tab(source, new Band(source, this.changed), saved);
    source.onProgress = () => this.changed();
    source.onOffer = (offer) => {
      t.offer = offer;
      this.offered(t);
    };
    return t;
  }

  /**
   * remove takes a source out, and its edits out of the log. The engine keeps
   * the last one, since a workspace of none has nothing to show or save.
   */
  async remove(tab: Tab): Promise<void> {
    await this.engine.remove(tab.source);
    const i = this.tabs.indexOf(tab);
    if (i < 0) return;
    this.tabs.splice(i, 1);
    this.relinked.delete(tab.id);
    if (this.showing === tab) this.showing = this.tabs[Math.min(i, this.tabs.length - 1)]!;
  }

  /**
   * relink points a tab at a file: one whose file has gone, or one whose file
   * changed under the log.
   *
   * The tab is replaced rather than repaired, because its columns, its rows and
   * its band all belong to the file behind it. What it keeps is its place in the
   * strip, the cell it was left on, and what the last save held, so nothing
   * about the session moves under the person doing it.
   */
  async relink(tab: Tab, ref: SourceRef): Promise<Tab> {
    const source = await this.engine.relink(tab.source, ref);
    const i = this.tabs.indexOf(tab);
    if (i < 0) return tab;

    const fresh = this.tab(source, tab.savedLog);
    fresh.cell = tab.cell;
    this.tabs[i] = fresh;
    this.relinked.add(fresh.id);
    if (this.showing === tab) this.showing = fresh;
    return fresh;
  }

  /** The sources, in the order the strip shows them. */
  get sources(): readonly Tab[] {
    return this.tabs;
  }

  /** The tab showing now. */
  get active(): Tab {
    return this.showing;
  }

  /** show makes another tab the one the grid draws and every edit goes to. */
  show(tab: Tab): void {
    if (this.tabs.includes(tab)) this.showing = tab;
  }

  /** The tab `step` places along from the one showing, wrapping at either end. */
  beside(step: number): Tab {
    const n = this.tabs.length;
    return this.tabs[(((this.tabs.indexOf(this.showing) + step) % n) + n) % n]!;
  }

  get name(): string {
    return this.showing.name;
  }

  get rows(): Rows {
    return this.showing.band;
  }

  /** The recogniser's question about the source showing. */
  get offer(): Offer | null {
    return this.mode === "transform" ? this.showing.offer : null;
  }

  get editable(): boolean {
    return this.mode === "transform";
  }

  transform(): void {
    this.mode = "transform";
    this.engine.mode(true);
  }

  /** view locks editing again. The log and the dirty dots stay. */
  view(): void {
    this.mode = "view";
    for (const t of this.tabs) t.offer = null;
    this.engine.mode(false);
  }

  /**
   * set types a value into one cell. It shows at once, the engine records it,
   * and a value the engine refuses is put back.
   */
  async set(row: number, col: number, value: string): Promise<void> {
    const t = this.showing;
    const pending = t.band.write(row, col, value);
    try {
      t.landed(await t.source.edit({ op: Op.Set, row, col, now: value }));
    } catch (err) {
      t.band.restore(pending);
      throw err;
    }
    t.band.settle(pending);
  }

  /** apply runs an offered program over its column: one edit, however long the column. */
  async apply(offer: Offer): Promise<void> {
    const t = this.tabs.find((tab) => tab.id === offer.source);
    if (t === undefined) throw new Error("the source that offer was about is gone");
    t.offer = null;
    t.landed(
      await t.source.edit({ op: Op.Apply, row: NO_ROW, col: offer.col, now: offer.program }),
    );
  }

  /**
   * undo takes the showing source's last edit back and returns it, so the cell
   * it changed can be shown. The engine replays the rest; no row is read again.
   */
  async undo(): Promise<Edit> {
    const t = this.showing;
    const changed = await t.source.undo();
    t.undone(changed);
    return changed.edit;
  }

  /** redo records again the edit undo last took back, and returns it. */
  async redo(): Promise<Edit> {
    const t = this.showing;
    const changed = await t.source.redo();
    t.landed(changed);
    return changed.edit;
  }

  /** find asks the engine for the next matching row in a column, read from the file. */
  find(req: FindRequest): Promise<Found> {
    return this.showing.source.find(req);
  }

  close(): void {
    this.engine.close();
  }

  /** What a save without a path should suggest: named after the first source. */
  get suggestedFileName(): string {
    const base = (this.tabs[0]?.name ?? "").replace(/\.[^.]*$/, "");
    return `${base || "workspace"}.uno`;
  }

  /** Whether anything would be lost by closing: an edit, a source added or
   * removed, or a source pointed at another file. */
  get dirty(): boolean {
    const ids = this.tabs.map((t) => t.id);
    const same =
      ids.length === this.savedSources.length && ids.every((id, i) => id === this.savedSources[i]);
    return !same || this.relinked.size > 0 || this.tabs.some((t) => t.dirty);
  }

  /** Whether a tab holds something the last save did not. */
  unsaved(tab: Tab): boolean {
    return tab.dirty || this.relinked.has(tab.id) || !this.savedSources.includes(tab.id);
  }

  /**
   * bytes asks the engine for the workspace as a .uno. `cell` is where the
   * grid is on the tab showing; every other tab is where it was left.
   *
   * `at` is where the file is going, which the writer needs before it writes:
   * a source under the same folder is pointed at relative to it, so the folder
   * can be copied somewhere else whole. That is why Save As asks for the path
   * first and serialises second.
   */
  bytes(cell: Cell, at: string): Promise<Uint8Array> {
    this.showing.cell = cell;
    return this.engine.save(
      {
        source: this.showing.id,
        cells: this.tabs.map((t) => ({ source: t.id, row: t.cell.row, col: t.cell.col })),
        at,
      },
      CARRY_LIMIT,
    );
  }

  /** Called once a save has landed, so the workspace stops reading as dirty. */
  saved(path: string): void {
    this.path = path;
    this.savedSources = this.tabs.map((t) => t.id);
    this.relinked.clear();
    for (const t of this.tabs) t.saved();
  }

  /** How far the index has read the showing file, as a whole percent. */
  indexed(): number {
    const p = this.showing.source.progress;
    return Math.floor((p.done / Math.max(1, p.total)) * 100);
  }

  /** What the status bar reports about the file showing. */
  status(): string {
    const t = this.showing;
    const p = t.source.progress;

    // A tab with no file behind it has no rows, no columns and no encoding to
    // report. What it has is a path that stopped working, which is the only
    // thing worth saying about it.
    if (t.missing) return `${t.trouble} · point it at a file to see its rows`;

    // Until the index reaches the end, the count is projected from how far it
    // has got, and says so.
    const parts = [
      `${p.complete ? "" : "≈"}${p.rows.toLocaleString()} rows`,
      `${t.band.cols()} columns`,
      t.source.opened.label,
    ];
    if (!p.complete) parts.push(`indexing ${this.indexed()}%`);

    const edits = t.edited;
    if (edits > 0) parts.push(`${edits} ${edits === 1 ? "edit" : "edits"}`);
    return parts.join(" · ");
  }
}
